src/com/facebook/buck/shell/DefaultWorkerTool.java

```diff
@@ -85,7 +103,47 @@ public int getMaxWorkers() {
   }
 
   @Override
+  public boolean isPersistent() {
+    return isPersistent;
+  }
+
+  @Override
   public ImmutableSortedSet<BuildRule> getRuntimeDeps() {
     return getDeps();
   }
+
+  @Override
+  public HashCode getInstanceKey() {
+    return buildOutputInitializer.getBuildOutput().getRuleKey().getHashCode();
+  }
+
+  @Override
+  public Data initializeFromDisk(OnDiskBuildInfo onDiskBuildInfo) throws IOException {
+    Optional<RuleKey> ruleKey =
+        onDiskBuildInfo.getRuleKey(BuildInfo.METADATA_KEY_FOR_RULE_KEY);
```

Comment:
This seems to be a unified way to do something regardless whether a rule has been built or restored from cache (i.e. when steps aren't run). At the point `initializeFromDisk` is called, `OnDiskBuildInfo` actually has `getRuleKey` called on it quite a few times, so I think this is actually faster than rehashing the key via the factory. If not - seems like `OnDiskBuildInfo` could use a cache.